### ISSUE TITLE
A pod cannot connect to itself via a service IP ("hairpin")

### ISSUE THREAD

**Issue by danwinship** [ovn-org/ovn-kubernetes#817]
A pod should be able to connect to a service IP that redirects back to itself, but this doesn't currently work.

eg, write this to a file, do `kubectl create -f FILE`:

    apiVersion: v1
    kind: Service
    metadata:
      name: test-service
    spec:
      selector:
        name: test
      ports:
      - port: 8080
    ---
    apiVersion: v1
    kind: Pod
    metadata:
      name: test-pod
      labels:
        name: test
    spec:
      containers:
      - name: test-container
        image: bmeng/hello-openshift
        ports:
        - containerPort: 8080
    ---
    apiVersion: v1
    kind: Pod
    metadata:
      name: test-pod-2
    spec:
      containers:
      - name: test-container
        image: bmeng/hello-openshift

then this will work:

    oc exec test-pod-2 -- curl -s test-service:8080

but this fails:

    oc exec test-pod -- curl -s test-service:8080

**Comment by girishmg** [comment]
I check the flow using `ovn-trace`, and it shows that the packet does get hair-pinned and is delivered back to the originating logical switch port. So, looks like things in OVN is fine. Perhaps, we need to change the `net.ipv4.conf.rp_filter` value in the pod to receive 'hair-pin' packets?

**Comment by danwinship** [comment]
> Perhaps, we need to change the `net.ipv4.conf.rp_filter` value in the pod to receive 'hair-pin' packets?

That doesn't seem to have any effect; tcpdump on either side of the veth doesn't show a return packet.

**Comment by girishmg** [comment]
Ok, this is the same issue as in #758 (where the traffic is hair-pinned back on management port -- the traffic from the host network pod in the master node towards the kubernetes service cluster ip is hairpinned back to the host network pod itself using management port).

In the ovs-vswitchd.log, I am seeing the following error:

```
2019-08-23T21:27:06.812Z|00002|dpif(handler1395)|WARN|system@ovs-system: execute ct(commit,zone=18,label=0/0x1),7 failed (Invalid argument) on packet tcp,vlan_tci=0x0000,dl_src=00:00:00:7c:87:25,dl_dst=0a:00:00:a8:00:06,nw_src=192.168.0.5,nw_dst=192.168.0.5,nw_tos=0,nw_ecn=0,nw_ttl=63,tp_src=9995,tp_dst=8080,tcp_flags=syn tcp_csum:b48b
```

The failure is in net/openswitch/conntrack.c`ovs_ct_execute() method and we are unable to perform the CT action.

`conntrack -E` showed the following incomplete conntrack transaction

    [NEW] tcp      6 120 SYN_SENT src=192.168.0.5 dst=10.102.166.105 sport=9995 dport=8080 [UNREPLIED] src=192.168.0.5 dst=192.168.0.5 sport=8080 dport=9995 zone=18

**Comment by squeed** [comment]
We may be able to fix this on the node with some iptables rules that cause hairpin traffic to be masqueraded from the node IP. Ugly, but I think other network plugins do something similar.

**Comment by girishmg** [comment]
@squeed I tried doing that but it didn't work for me. Please see https://github.com/ovn-org/ovn-kubernetes/issues/758#issuecomment-538015743 for more information.

For the issue captured here, the traffic starts from the pod, gets hair pinned by the OVN pipeline, and  is forwarded back into the pod. So, the only place where these packets are visible in the node is on the port on br-int that connects to the pod. Are you suggesting that we had iptable rules on those OVS ports?

**Comment by dceara** [comment]
I've been looking at this issue lately and there are a few problems:
1. with the current OVN code when the packet is hairpinned we end up trying to commit the same conntrack entry twice in the pod's logical port conntrack zone (`src=pod-ip, dst=vip; reply-src=pod-ip, reply-dst=pod-ip`). The kernel refuses to do that and increments the `insert_failed` conntrack counter; the packet is then dropped.
2. even if the kernel would allow the second `ct_commit`, we'd still have an issue because the SYN packet reaching the pod will have `src=pod-ip, dst=pod-ip` so the kernel will reply locally instead of sending the SYN-ACK through OVN. The connection would fail as the TCP client in SYN-SENT is created for `src=pod-ip, dst=VIP`.

I think to address both issues we have two options, both of which force replies to come back through OVN:
1. either configuring all load balancers on the the gateway routers where we can use `lb_force_snat_ip` to change the source IP for load balanced traffic. This might not be so straightforward though.
2. handle hairpinning scenarios in core OVN. I've sent a patch series for that with which hairpinned traffic is automatically SNAT-ed using the VIP as new source IP: https://patchwork.ozlabs.org/project/openvswitch/list/?series=153591. If/once it gets accepted this hairpin issue should be resolved.

**Comment by girishmg** [comment]
@dceara this fix still doesn't support the Hairpin issue on the management port that we discussed during the OVS/OVN Conference in the November, right? https://github.com/ovn-org/ovn-kubernetes/issues/758

**Comment by dceara** [comment]
> @dceara this fix still doesn't support the Hairpin issue on the management port that we discussed during the OVS/OVN Conference in the November, right? #758

@girishmg No, we still don't have an OVN-only solution for that.

**Comment by danwinship** [comment]
This is fixed by OVN 2.13
